packages/editor/src/components/main-element-warnings/index.js

```diff
@@ -0,0 +1,57 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { useEffect } from '@wordpress/element';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+import { store as noticesStore } from '@wordpress/notices';
+
+/**
+ * Internal dependencies
+ */
+import { store as editorStore } from '../../store';
+
+const checkMainTag = ( blocks, mainTagCount ) => {
+	blocks.forEach( ( block ) => {
+		if ( block.attributes.tagName === 'main' ) {
+			mainTagCount++;
+		}
+		// If the block has innerBlocks, call the function again.
+		if ( block.innerBlocks.length > 0 ) {
+			mainTagCount = checkMainTag( block.innerBlocks, mainTagCount );
+		}
+	} );
+
+	return mainTagCount;
+};
+
+export default function MainElementWarnings() {
+	const { type, blocks } = useSelect( ( select ) => {
+		const postType = select( editorStore ).getCurrentPostType();
+		const { getBlocks } = select( blockEditorStore );
+		return {
+			type: postType,
+			blocks: getBlocks(),
```

Comment:
We could move the post type check here and only return the `blocks` value when editing a template. This will avoid unnecessary tracking for other post types.